### ISSUE TITLE
Request for Vertical (Stacked) Editor Tabs Display Option in DBeaver

### ISSUE THREAD

**Issue by shaochangxing** [dbeaver/dbeaver#23248]
Hello DBeaver Team,

I am a frequent user of DBeaver and really appreciate the functionality it offers. However, I've been looking for a way to customize the layout of query tabs within the application to display vertically instead of horizontally, which would be particularly useful when working with numerous open queries or limited screen real estate.

Currently, it seems that DBeaver doesn't natively support a vertical orientation for its query tabs. This feature is often found in other IDEs and could greatly enhance the user experience by allowing more efficient navigation among multiple queries.

Could you please consider adding an option to configure the query tabs to stack vertically? It would help users who work on high-resolution monitors or have multi-monitor setups where space management is crucial.

Thank you very much for your consideration, and I look forward to any updates or solutions regarding this request.

**Comment by E1izabeth** [comment]
Hello, @shaochangxing 
Please take a look at [this feature.](https://github.com/dbeaver/dbeaver/wiki/SQL-Editor#multiple-results-in-one-tab). It sounds like what you want.

**Comment by dataract** [comment]
> Hello, @shaochangxing Please take a look at [this feature.](https://github.com/dbeaver/dbeaver/wiki/SQL-Editor#multiple-results-in-one-tab). It sounds like what you want.

I don't believe this suggestion is what we need.

**Comment by dataract** [comment]
I was looking for this same resource that @shaochangxing is asking for. (idea/example below)
![Screenshot_2024-03-28_150509](https://github.com/dbeaver/dbeaver/assets/161634184/4ec55388-a8f0-4e7e-8a7f-d816d8620dc0)

**Comment by E1izabeth** [comment]
> I don't believe this suggestion is what we need.

We would greatly appreciate it, if you could explain why.

**Comment by shaochangxing** [comment]
@E1izabeth 
![1](https://github.com/dbeaver/dbeaver/assets/11647361/713793a9-b247-4b92-95a9-7f61e32e2254)
The picture below shows what I want
![2](https://github.com/dbeaver/dbeaver/assets/11647361/58597939-be4e-4cf4-92f8-ea1d7cf37356)

**Comment by humphreygao** [comment]
改tab比较麻烦，要到上游eclipse platform

**Comment by shaochangxing** [comment]
> 改tab比较麻烦，要到上游eclipse platform

那有点难受，我经常要打开不同数据库的十几个表，切换起来有点麻烦。

**Comment by humphreygao** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=58945


https://git.eclipse.org/r/c/platform/eclipse.platform.swt/+/173778/14#message-396057146457f282afe0e4a2b76c01d9fb505022

代码满足不了要求，后面没人推进了

**Comment by dataract** [comment]
> > I don't believe this suggestion is what we need.
> 
> We would greatly appreciate it, if you could explain why.
I posted the idea here...
https://github.com/dbeaver/dbeaver/issues/23248#issuecomment-2025810640
there isn't enough space for horizontal tabs and became hard do navigate/visualize because they are hidden (off screen)

**Comment by E1izabeth** [comment]
Thank you for explanation. It's a duplicate of #14546, so I'm closing this issue.
You can vote for the mentioned issue with 👍 under the issue description to promote it to a higher priority.

**Comment by dataract** [comment]
The original poster there is only asking for the result tabs. We also want  the query/script tabs to be arrranged vertically.

**Comment by dataract** [comment]
> Thank you for explanation. It's a duplicate of #14546, so I'm closing this issue. You can vote for the mentioned issue with 👍 under the issue description to promote it to a higher priority.

The original poster there is only asking for the result tabs. We also want the query/script tabs to be vertically arranged.

**Comment by E1izabeth** [comment]
OK
So, for clarification: the issue is about placing SQL Editor (and probably Entity editor) tabs vertically.

**Comment by dataract** [comment]
> OK So, for clarification: the issue is about placing SQL Editor (and probably Entity editor) tabs vertically.

Yes (SQL Editor). Me and @shaochangxing placed examples (screenshots) displaying our examples/ideas that work in the way we explained.

**Comment by sethleonard** [comment]
Agreed on the need for vertical tabs. I would also add the alternative of allowing for customizing of the tab labels; when they all start with the current database name, that's all you see when you have multiple tabs open.

<img width="1089" alt="image" src="https://github.com/user-attachments/assets/881c5a2a-11dc-4c7b-a10a-513e69c6793a">

**Comment by REYJL** [comment]
One way would be to have a flag in "Project" panel where scripts are displayed vertically.
Could be:

- Not open = No flag
- Opened but not modified = green check
- Opened and unsaved changes = *

Just a suggestion

tha said, I would like vertical stack of tabs (position able) 
![image](https://github.com/user-attachments/assets/3e57051b-fb2a-416b-a7a9-10f8ca7b54e6)

**Comment by DavidHJong** [comment]
Same request. It's very convenient when many tabs are opened. 

For clarity, refer to Microsoft Edge vertical tabs.

<img width="1280" alt="Screenshot 2024-10-15 165630" src="https://github.com/user-attachments/assets/291be81e-f37b-4b7b-83ff-58e541581b29">

**Comment by chesini** [comment]
> Agreed on the need for vertical tabs. I would also add the alternative of allowing for customizing of the tab labels; when they all start with the current database name, that's all you see when you have multiple tabs open.
> <img alt="image" width="1089" src="https://private-user-images.githubusercontent.com/20046237/366609357-881c5a2a-11dc-4c7b-a10a-513e69c6793a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNDQ0MjAsIm5iZiI6MTczNjM0NDEyMCwicGF0aCI6Ii8yMDA0NjIzNy8zNjY2MDkzNTctODgxYzVhMmEtMTFkYy00YzdiLWExMGEtNTEzZTY5YzY3OTNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEzNDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ZmViY2MwZjZmYmNiYTFhYzQxOThmODA1NDVkZjc5OGNmYTA4ZDEzNjQzMGQzOTg0ZWYyMGU1MDMwYzg2MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iyXU6zrMiP_5v6f5NtVMlQLQi_9r-38-e5W1-6x84xY">

Maybe this: https://github.com/dbeaver/dbeaver/issues/9708#issuecomment-691033526 could help you with this script tabs title issue but is only applied to the script files.

**Comment by chesini** [comment]
> Agreed on the need for vertical tabs. I would also add the alternative of allowing for customizing of the tab labels; when they all start with the current database name, that's all you see when you have multiple tabs open.
> <img alt="image" width="1089" src="https://private-user-images.githubusercontent.com/20046237/366609357-881c5a2a-11dc-4c7b-a10a-513e69c6793a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNDQ0MjAsIm5iZiI6MTczNjM0NDEyMCwicGF0aCI6Ii8yMDA0NjIzNy8zNjY2MDkzNTctODgxYzVhMmEtMTFkYy00YzdiLWExMGEtNTEzZTY5YzY3OTNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEzNDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ZmViY2MwZjZmYmNiYTFhYzQxOThmODA1NDVkZjc5OGNmYTA4ZDEzNjQzMGQzOTg0ZWYyMGU1MDMwYzg2MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iyXU6zrMiP_5v6f5NtVMlQLQi_9r-38-e5W1-6x84xY">

Maybe this: https://github.com/dbeaver/dbeaver/issues/9708#issuecomment-691033526 could help you @sethleonard with this script tabs title issue.

**Comment by sethleonard** [comment]
> > Agreed on the need for vertical tabs. I would also add the alternative of allowing for customizing of the tab labels; when they all start with the current database name, that's all you see when you have multiple tabs open.
> > <img alt="image" width="1089" src="https://private-user-images.githubusercontent.com/20046237/366609357-881c5a2a-11dc-4c7b-a10a-513e69c6793a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNDQ0MjAsIm5iZiI6MTczNjM0NDEyMCwicGF0aCI6Ii8yMDA0NjIzNy8zNjY2MDkzNTctODgxYzVhMmEtMTFkYy00YzdiLWExMGEtNTEzZTY5YzY3OTNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEzNDg0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0ZmViY2MwZjZmYmNiYTFhYzQxOThmODA1NDVkZjc5OGNmYTA4ZDEzNjQzMGQzOTg0ZWYyMGU1MDMwYzg2MmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iyXU6zrMiP_5v6f5NtVMlQLQi_9r-38-e5W1-6x84xY">
> 
> Maybe this: [#9708 (comment)](https://github.com/dbeaver/dbeaver/issues/9708#issuecomment-691033526) could help you @sethleonard with this script tabs title issue.

Perfect. Thank you, @chesini !

**Comment by DavidHJong** [comment]
@humphreygao is right. Right now, if we're looking to add vertical (or stacked) editor tabs in DBeaver, we'll run into a fundamental limitation with the Eclipse platform, which DBeaver is built on. The problem is that Eclipse's UI toolkit (SWT) uses a component called CTabFolder, and it doesn't support vertical or multi-row tabs at all. This issue has been on the radar since at least 2004, as seen in Eclipse Bug 58945, but no one has managed to fix it yet.

Because DBeaver relies on SWT for its interface, getting vertical tabs would mean having to change the underlying CTabFolder widget in SWT. That’s not a small task—it would involve contributing directly to the Eclipse platform, following their development guidelines, and committing to long-term maintenance of those changes.

An alternative would be to create a custom tab management system within DBeaver to mimic vertical tabs. However, this workaround would require a lot of development effort and might not mesh perfectly with the existing SWT framework.

In summary, adding vertical editor tabs in DBeaver depends on overcoming deep-rooted limitations in the Eclipse platform. Our best bet would be to either contribute to the Eclipse SWT project or develop a creative workaround within DBeaver, though both paths come with significant challenges.

**Comment by DavidHJong** [comment]
Currently when many tabs are open, clicking the right arrow reveals a dropdown menu. Could we instead make this dropdown permanently visible--pinned as a right-hand pane--to serve as a vertical tab list? Is that feasible? 

 ![Image](https://github.com/user-attachments/assets/b14528bb-35e3-4093-a50a-caa242936832)

**Comment by DavidHJong** [comment]
Currently when many tabs are open, clicking the right arrow reveals a dropdown menu. Could we instead make this dropdown permanently visible--pinned as a right-hand pane--to serve as a vertical tab list? Is that feasible?

![Image](https://github.com/user-attachments/assets/b36dc5fe-940f-4e25-973e-b4d90d0bc210)

**Comment by iron0012** [comment]
That solution would be great, except the order of the tabs in that dropdown menu appears to be random, or else follows some ordering rule which is not apparent to the user. There is another issue about this. The order of the tab names in that pull-down should be static and follow the order in which they were spawned. Instead, it appears to be random and shuffles over time.

Here's the other pertinent issue https://github.com/dbeaver/dbeaver/issues/5400

(also @DavidHJong if I'm understanding the issue correctly your screenshot is not quite correct, since we're talking about the open SQL editor tabs, and the pulldown menu for the open editor tabs which cannot fit on screen, not the columns in the results tab)

**Comment by DavidHJong** [comment]
You make a good point, @iron0012. I originally interpreted this post as requesting vertical arrangement of all tabs, not just query tabs specifically. In my opinion, arranging only query tabs offers limited benefit when working with multiple schemas and tables simultaneously. Since query tabs function similarly to schema and table tabs in the interface, my screenshot demonstrates the same concept, even though it contains numerous column tabs. What do you think?

**Comment by mstaal** [comment]
Any progress on this one?

**Comment by E1izabeth** [comment]
It's not planned for the nearest releases.
You can vote for the issue with 👍 under the issue description to promote it to a higher priority.
